### PR TITLE
Inbound queue v2

### PR DESCRIPTION
**Pull request by claravanstaden**
Resolves: [SNO-1164](https://linear.app/snowfork/issue/SNO-1164)

### CODE REVIEW

**Comment by claravanstaden** [comment]
@yrong would appreciate an early review here. 🙏🏻

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 1]
```diff
@@ -2,454 +2,170 @@
...
 
-		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
-		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
-		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+		let network = EthereumNetwork::get();
 
+		let fee_asset = Location::new(1, Here);
+		let fee_value = 1_000_000_000u128; // TODO needs to be dry-run to get the fee but also
+									 // need to add a fee here for the dry run... Chicken/egg problem?
+		let fee: Asset = (fee_asset, fee_value).into();
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs]: IIUC the `dry_run` on BH is to convert the InboundMessage to a XCM.  Then UI can dry run that XCM on AH for the cost.
 
So IMO `InboundMessage` should contain the fee like V1 before,  it can be zero for estimation, after that filled in with the result for calling the Gateway contract from UX.

File: bridges/snowbridge/primitives/router/src/inbound/dry_run.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			xcm_program.try_into().map_err(|error| DryRunError::VersionedConversionFailed)?;
+		let mut hash = xcm.using_encoded(sp_io::hashing::blake2_256);
+		frame_system::Pallet::<Runtime>::reset_events(); // To make sure we only record events from current call.
+		let result = XcmExecutor::prepare_and_execute(
+			origin_location,
+			xcm,
+			&mut hash,
+			Weight::MAX, // Max limit available for execution.
+			Weight::zero(),
+		);
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/dry_run.rs]: It's not to execute the xcm on BH, supposed to send from BH and executed on AH.


https://github.com/claravanstaden/polkadot-sdk/pull/3/files#r1834062480

 Once we got the xcm on AH(fee can be zero initially), these apis can be used to estimate the real cost.


image

image

image

https://github.com/paritytech/polkadot-sdk/pull/3872 FYI

Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/dry_run.rs]: Got it, thanks @yrong!

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 1 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs]: @yrong do you think the AH execution fee (which can be 0 for estimation) should be added as a field in [MessageV2](https://github.com/claravanstaden/polkadot-sdk/pull/3/files#diff-011a2c280539a541db601aa2fd4e5f30960525c4bb35588c3de78effb01720fcR29)?

Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs]: @vgeddes or how did you imagine the AH execution fee be specified? I see there is a fee added for register token (which is the foreign asset creation deposit) but no other place in the contracts to specify the AH execution fee. Iiuc this will be the flow:

- Inbound queue runtime API to convert command to XCM
- UX takes XCM and dry-runs the XCM on AH to get fee
- UX takes fee number and adds it to the command, emits OutboundEvent

I don't see such a fee parameter in the contract code, so just wanna get your thoughts on this.

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 3]
```diff
@@ -244,63 +195,35 @@ pub mod pallet {
...
-				xcm,
-				fee
-			);
+			// Todo: Deposit fee(in Ether) to RewardLeger which should cover all of:
+			// T::RewardLeger::deposit(who, envelope.fee.into())?;
+			// a. The submit extrinsic cost on BH
+			// b. The delivery cost to AH
+			// c. The execution cost on AH
+			// d. The execution cost on destination chain(if any)
+			// e. The reward
```
Comment (claravanstaden) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:214]: @yrong just confirming, this will be calculated off-chain, right?

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 4]
```diff
@@ -215,9 +166,9 @@ pub mod pallet {
 		}
 	}
 
-	/// The current nonce for each channel
+	/// The nonce of the message been processed or not
 	#[pallet::storage]
-	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, ChannelId, u64, ValueQuery>;
+	pub type Nonce<T: Config> = StorageMap<_, Identity, u64, bool, ValueQuery>;
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: The `Identity` hasher isn't safe in this context. See https://docs.substrate.io/build/runtime-storage/#hashing-algorithms for reasons:

```suggestion
	pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, u64, bool, ValueQuery>;
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 5]
```diff
@@ -230,7 +181,7 @@ pub mod pallet {
 		#[pallet::call_index(0)]
 		#[pallet::weight(T::WeightInfo::submit())]
 		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
-			let who = ensure_signed(origin)?;
+			let _who = ensure_signed(origin)?;
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:184]: Is this simplification not possible?
```suggestion
			ensure_signed(origin)?;
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 4 above)
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: The idiomatic way to implement a _Set_ data structure using storage maps is to use the empty value `()` as the storage value:
```suggestion
	pub type Nonce<T: Config> = StorageMap<_, Identity, u64, (), ValueQuery>;
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 6]
```diff
@@ -244,63 +195,35 @@ pub mod pallet {
...
-				&sovereign_account,
-				Preservation::Preserve,
-				Fortitude::Polite,
-			)
-			.min(delivery_cost);
-			if !amount.is_zero() {
-				T::Token::transfer(&sovereign_account, &who, amount, Preservation::Preserve)?;
-			}
+			// Verify the message has not been processed
+			ensure!(!<Nonce<T>>::contains_key(envelope.nonce), Error::<T>::InvalidNonce);
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs]: Prefer this notation for referencing storage fields
```suggestion
			ensure!(!Nonce::<T>::contains_key(envelope.nonce), Error::<T>::InvalidNonce);
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 7]
```diff
@@ -244,63 +195,35 @@ pub mod pallet {
...
 
-			// Decode payload into `VersionedMessage`
-			let message = VersionedMessage::decode_all(&mut envelope.payload.as_ref())
+			// Decode payload into `MessageV2`
+			let message = MessageV2::decode_all(&mut envelope.payload.as_ref())
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:202]: Lets make use of rust modules as namespaces, instead of appending `V2` to type names.

```suggestion
			let message = Message::decode_all(&mut envelope.payload.as_ref())
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 3 above)
Comment (vgeddes) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:214]: yeah 👍

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 8]
```diff
@@ -244,63 +195,35 @@ pub mod pallet {
...
-				channel_id: envelope.channel_id,
-				nonce: envelope.nonce,
-				message_id,
-				fee_burned: fee,
-			});
+			// Set nonce flag to true
+			<Nonce<T>>::try_mutate(envelope.nonce, |done| -> DispatchResult {
+				*done = true;
+				Ok(())
+			})?;
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs]: Much simpler:
```suggestion
			Nonce::<T>::insert(envelope.nonce, ())
```

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 9]
```diff
@@ -318,61 +241,4 @@ pub mod pallet {
...
-		pub fn calculate_delivery_cost(length: u32) -> BalanceOf<T> {
-			let weight_fee = T::WeightToFee::weight_to_fee(&T::WeightInfo::submit());
-			let len_fee = T::LengthToFee::weight_to_fee(&Weight::from_parts(length as u64, 0));
-			weight_fee
-				.saturating_add(len_fee)
-				.saturating_add(T::PricingParameters::get().rewards.local)
-		}
-
-		/// Burn the amount of the fee embedded into the XCM for teleports
-		pub fn burn_fees(para_id: ParaId, fee: BalanceOf<T>) -> DispatchResult {
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:347]: We still need to teleport some DOT from the relayers account to AH so it can be used to pay for the execution of the first few instructions in the XCM.

Referring back to this slack [convo](https://snowfork.slack.com/archives/D02U9P3G3B5/p1730820291303089?thread_ts=1730800779.986149&cid=D02U9P3G3B5), enough DOT needs to be teleported to AH to pay for execution of `prologue`.

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs [hunk 10]
```diff
@@ -244,63 +195,35 @@ pub mod pallet {
...
+			// b. The delivery cost to AH
+			// c. The execution cost on AH
+			// d. The execution cost on destination chain(if any)
+			// e. The reward
 
-			// Burning fees for teleport
-			Self::burn_fees(channel.para_id, fee)?;
+			// Attempt to forward XCM to AH
+			let dest = Location::new(1, [Parachain(T::AssetHubParaId::get())]);
+			let (message_id, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```
Comment (vgeddes) [inline bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:218]: This `send_xcm` helper seems to be obsolete and is too low-level. Can you try using this `SendController` interface instead: https://paritytech.github.io/polkadot-sdk/master/staging_xcm_builder/trait.SendController.html

It also handles XCM version management appropriately.

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 11]
```diff
@@ -2,454 +2,172 @@
...
+};
 
-const MINIMUM_DEPOSIT: u128 = 1;
+const LOG_TARGET: &str = "snowbridge-router-primitives";
 
 /// Messages from Ethereum are versioned. This is because in future,
 /// we may want to evolve the protocol so that the ethereum side sends XCM messages directly.
 /// Instead having BridgeHub transcode the messages into XCM.
 #[derive(Clone, Encode, Decode, RuntimeDebug)]
 pub enum VersionedMessage {
```
Comment (vgeddes) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs:24]: This versioned type is unused in V2, in the sense that the ethereum contracts do not send it.

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 12]
```diff
@@ -2,454 +2,172 @@
...
-	V1(MessageV1),
+	V2(Message),
 }
 
-/// For V1, the ethereum side sends messages which are transcoded into XCM. These messages are
+/// For V2, the ethereum side sends messages which are transcoded into XCM. These messages are
```
Comment (vgeddes) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs]:
```suggestion
/// The ethereum side sends messages which are transcoded into XCM on BH. These messages are
```

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 13]
```diff
@@ -2,454 +2,172 @@
...
-	/// The funds will deposited into the sovereign account of destination parachain `para_id` on
-	/// AssetHub, Account `id` on the destination parachain will receive the funds via a
-	/// reserve-backed transfer. See <https://github.com/paritytech/xcm-format#depositreserveasset>
-	ForeignAccountId32 {
-		para_id: u32,
-		id: [u8; 32],
-		/// XCM execution fee on final destination
-		fee: u128,
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum InboundAsset {
```
Comment (vgeddes) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs]: Prefer not to prefix with `Inbound`.

```suggestion
pub enum Asset {
```

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 14]
```diff
@@ -2,454 +2,172 @@
...
 
+		let fee_asset = Location::new(1, Here);
+		let fee_value = 1_000_000_000u128; // TODO get from command
+		let fee: Asset = (fee_asset, fee_value).into();
 		let mut instructions = vec![
-			ReceiveTeleportedAsset(total_fee_asset.into()),
-			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
-			DescendOrigin(PalletInstance(inbound_queue_pallet_index).into()),
+			ReceiveTeleportedAsset(fee.clone().into()),
+			BuyExecution { fees: fee, weight_limit: Unlimited },
```
Comment (vgeddes) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs:118]: Should rather use the new `PayFees` instruction that is in XCMv5

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs [hunk 15]
```diff
@@ -2,454 +2,172 @@
...
+			.push_interior(AccountKey20 { key: message.origin.into(), network: None })
+			.map_err(|_| ConvertMessageError::InvalidXCM)?;
 
-		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
-		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
-		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+		let network = EthereumNetwork::get();
 
+		let fee_asset = Location::new(1, Here);
+		let fee_value = 1_000_000_000u128; // TODO get from command
```
Comment (vgeddes) [inline bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: As I noted in a previous comment, this fee should also be burnt from the relayers account on BH.

Some other notes:
* Should make the fee configurable via runtime config, ie add `T::XcmPrologueFee: Balance` pallet config
* Add to your TODOs to implement an XCM simulation test that calculates an appropriate minimum fee, assuming the worst case where a maximum of eight assets are supplied. Then we add 2x buffer on top of that minimum.

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 4 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: @vgeddes on the topic of nonces, how would you feel about changing this storage item to only keep pending nonces, like here: https://github.com/yrong/polkadot-sdk/pull/4/files#diff-10cb33bd6056f3960af963e776896e440e942580843f1676c6d0be393ad61010R220 When a message is processed successfully, it is removed from the map?

This approach has 2 benefits:
- The storage map will not grow boundlessly as the nonce grows
- It is easier for relayers to query this map to see which messages should be relayed, instead of querying the whole map of all previous nonces relayed and picking out the outstanding ones.

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 5 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:184]: I kept the variable because it will be used for the rewards ledger once the unified rewards PR is merged: https://github.com/claravanstaden/polkadot-sdk/pull/3/files#diff-bfe761b371a553619edf5be2dd792c45bd14f1a916e51b77f5b77b9395d92cdaR209

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 15 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: > Should make the fee configurable via runtime config, ie add T::XcmPrologueFee: Balance pallet config

Can we make this fee configurable? Iirc we cannot, because it could include arbitrary instructions such as Transact. I explained the flow as I see it should work over here: https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1846293603

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 11 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:24]: @vgeddes I wonder why we don't version the message this time around, I made a comment about it here too: https://github.com/yrong/polkadot-sdk/pull/4#discussion_r1846402607

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 4 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: Nevermind, brain fart. The pallet needs to keep track of which nonces have been processed otherwise a message can be processed twice. 😅

Comment (vgeddes) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: The problem is that only the Ethereum gateway contract knows which outbound nonces are pending delivery to the inbound queue pallet on BH.

Actually I remember now that I originally wanted to use a sparse bitmap structure to make this nonce tracking more space efficient

ie.

```
pub type Nonce<T: Config> = StorageMap<_, Twox64Concat, u128, u128, ValueQuery>
```

Each "bucket" can track 128 sequential nonces.

I do this for the Ethereum side, where the delivered inbound nonces are tracked in this data structure: https://github.com/Snowfork/snowbridge/blob/553b0b9d9ec6f3683c6c605417ea5532282771ed/contracts/src/utils/SparseBitmap.sol

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 7 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:202]: It is actually just called Message, but aliased to MessageV2 here: https://github.com/claravanstaden/polkadot-sdk/pull/3/files#diff-bfe761b371a553619edf5be2dd792c45bd14f1a916e51b77f5b77b9395d92cdaR52 since I use `Message` from `snowbridge_core` too.

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 15 above)
Comment (yrong) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: > Can we make this fee configurable? Iirc we cannot, because it could include arbitrary instructions such as Transact. I explained the flow as I see it should work over here: https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1846293603

Yeah, I still think InboundMessage should contain the fee like V1 before with AssetId including.

But there is also another option to make it configurable, A Nmap storage item on BH something like:

| Destination | Operation | FeeId | FeeAmount |
| -------- | ------- | -------- | ------- |
| AH      | TransferENA | WETH | 0.0001 |
| AH      | TransferENA | DOT | 0.01 |
| AH      | TransferPNA | DOT  | 0.01  |
| Moonbeam |   TransferENA | GLMR | 0.1 |
| Moonbeam |  Call_Contract_Method_1 | GLMR | 0.5   | 
| Moonbeam |  Call_Contract_Method_2 | GLMR | 10   | 

- Destination chain can config multiple assets as fee. e.g. for `TransferENA` to AH both WETH and DOT can be configured as fee.
-  For multiple hop operations the fee is total of the execution cost on AH and on destination chain. e.g. for `TransferENA` to Moonbeam the fee is 0.0001 WETH + 0.1 GLMR
- Operations with Destination to AH should be configured with governance call from Polkadot 
- Operations with Destination to Moonbeam should be configured with governance call from Moonbeam

IMO this is doer-able because
- fee on Substrate is very limited and invariable, comparing with the fee on Ethereum side, so `dry_run` every time may not be necessary
- we can add `SetAppendix(DepositToBeneficiary)` and return the extra fee to end user, so set a high fee safe enough is not a big deal

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 1 above)
Comment (yrong) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs]: Another option as mentioned in https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1851214454

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 10 above)
Comment (yrong) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:218]: Then we need to explicitly wrap xcm into a VersionXcm. I don't think this is necessary. 

AFAIK with `pallet-xcm` version discovery/negotiation is automatically scheduled 
https://github.com/paritytech/polkadot-sdk/blob/70b6c7b08a0bc27c6155bca66461af97b829d208/polkadot/xcm/pallet-xcm/src/lib.rs#L820

then send call will just pick up the right version 
https://github.com/paritytech/polkadot-sdk/blob/70b6c7b08a0bc27c6155bca66461af97b829d208/cumulus/pallets/xcmp-queue/src/lib.rs#L977

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 15 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: That's an interesting idea @yrong. It could work. I think the main downside to this (from my perspective) is that many governance calls will be a pain to do. It will require more invention from our team's side, more coordination with other teams.

What is your concern with the dry-run approach? Isn't runtime API calls free so there's no cost to doing the estimation?

Comment (yrong) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: > I still think InboundMessage should contain the fee like V1 before but with AssetId including.

Yeah, that's why I still prefer this.

Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: @vgeddes let me know what you think too.

**Comment by claravanstaden** [comment]
Closing in favour of https://github.com/claravanstaden/polkadot-sdk/pull/4. The diff in this PR included changes already made in another PR, so it was hard to see the true diff.

File: bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs (hunk 4 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: Added. 

SparseBitmap: https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-4ff1da0a1e9f3d7498050866dabf1fe730bd941380812d0e741ef5ba7fc031d3
Updated Nonce storage: https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-bfe761b371a553619edf5be2dd792c45bd14f1a916e51b77f5b77b9395d92cdaR174

Comment (claravanstaden) [reply bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs:171]: Now irrelevant with the sparse bitmap data structure: https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1851556972

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 13 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs]: https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-011a2c280539a541db601aa2fd4e5f30960525c4bb35588c3de78effb01720fcR43

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 14 above)
Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:118]: https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-011a2c280539a541db601aa2fd4e5f30960525c4bb35588c3de78effb01720fcR118

File: bridges/snowbridge/primitives/router/src/inbound/v2.rs (hunk 15 above)
Comment (vgeddes) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: I think there's a misunderstanding here how fees are supposed to be paid on AH. Please read again the [API docs](https://github.com/Snowfork/snowbridge/blob/553b0b9d9ec6f3683c6c605417ea5532282771ed/contracts/src/v2/IGateway.sol#L54) for `IGateway.sendMessage`:

>     // The `xcm` should contain the necessary instructions to:
>     // 1. Pay XCM execution fees for `xcm`, either from assets in holding,
>     //    or from the sovereign account of `msg.sender`.
>     // 2. Handle the assets in holding, either depositing them into
>     //    some account, or forwarding them to another destination.

Or in other words, its the users responsibility to provide assets and xcm instructions to cover fees. I.e the user-provided `xcm` can contain `PayFee` instructions.

Please let me know if there is anything that I should clarify about this.

Its only the fixed xcm _prologue_ that requires a small fixed fee to be teleported by the relayer account on BH. To cover the first few instructions which are prepending to the users `xcm`.

Our governance messages for registering ENA tokens will use the same API.

Comment (claravanstaden) [reply bridges/snowbridge/primitives/router/src/inbound/v2.rs:114]: I see, thanks for clearing it up @vgeddes. I will add it.
